packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,360 @@
+const DEFAULT_LOGGING_INTERVAL = 10_000;
+
+const log = debug("waku:connection-manager");
+
+export interface Libp2pComponents {
+  connectionManager: Libp2p["connectionManager"];
+  peerStore: Libp2p["peerStore"];
+  ping: Libp2p["ping"];
+  dial: Libp2p["dial"];
+}
+
+export interface UpdatedStates {
+  allPeers: Peer[];
+  allDialablePeers: Peer[];
+  bootstrapPeers: Peer[];
+  nonBootstrapPeers: Peer[];
+  dialableBootstrapPeers: Peer[];
+  dialableNonBootstrapPeers: Peer[];
+  allConnections: Connection[];
+  bootstrapConnections: Connection[];
+  nonBootstrapConnections: Connection[];
+}
+
+export interface Options {
+  /**
+   * Number of attempts before a peer is considered non-dialable
+   * This is used to not spam a peer with dial attempts when it is not dialable
+   */
+  maxDialAttemptsForPeer?: number;
+  /**
+   * Max number of bootstrap peers allowed to be connected to, initially
+   * This is used to increase intention of dialing non-bootstrap peers, found using other discovery mechanisms (like Peer Exchange)
+   * Is overridden by {@link maxDialAttemptsBeforeBootstrapFallback}
+   */
+  maxBootstrapPeersAllowed?: number;
+  /**
+   * Number of attempts before dialling bootstrap nodes over the {@link maxBootstrapPeersAllowed} value.
+   * Dialing is the process of opening an outgoing connection to a listening peer with a transport that is supported by both peers.
+   * This is only used when other discovery mechanisms haven't yield peers that are dialable
+   * This increases relative centralization of the network
+   */
+  maxTimeBeforeBootstrapFallbackMs?: number;
+}
+
+export class ConnectionManager extends KeepAliveManager {
+  private static instance: ConnectionManager;
+  private options: Options;
+  private libp2pComponents: Libp2pComponents;
+  private dialAttemptsForPeer: Map<string, number> = new Map();
+  private loggingCounter = 0;
+
+  public static create(
+    libp2p: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ): ConnectionManager {
+    if (!ConnectionManager.instance) {
+      ConnectionManager.instance = new ConnectionManager(
+        libp2p,
+        keepAliveOptions,
+        relay,
+        options
+      );
+    }
+    return ConnectionManager.instance;
+  }
+
+  constructor(
+    libp2pComponents: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ) {
+    super();
+    this.libp2pComponents = libp2pComponents;
+    this.options = options ?? {};
+
+    this.runService(keepAliveOptions, relay)
+      .then(() => log(`Connection Manager is now running`))
+      .catch((error) =>
+        log(`Unexpected error while running service: ${error}`)
+      );
+  }
+
+  private async runService(
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay
+  ): Promise<void> {
+    const {
+      maxTimeBeforeBootstrapFallbackMs = DEFAULT_APPROX_TIME_BEFORE_BOOTSTRAP_FALLBACK_MS,
```

Comment:
I see that you wait before connecting to more bootstrap nodes, in case we don't find enough peers via peer exchange.

Looking back at the discussion: https://github.com/waku-org/js-waku/issues/914#issuecomment-1372943986 this is not a logic we discussed.

I would suggest to not include this logic (yet).

Simply:
1. Connect to 1 bootstrap node
2. Find more peers using other discovery mechansims
3. Connect to all enw peers found until requirements are fullfilled

As you described here: https://github.com/waku-org/js-waku/issues/914#issuecomment-1376892344

I would strongly recommend to KISS this first PR by keep the logic barebone.

Reply:
> But we could add it as the next iteration for the same PR

I don't expect it because we don't want web client to hog connections on the bootstrap fleet. The solution is to improve peer exchange.